cpp/src/utilities/device_assert.cuh

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef DEVICE_ASSERT_CUH
+#define DEVICE_ASSERT_CUH
+
+#include <cuda_runtime.h>
+
+/**---------------------------------------------------------------------------*
+ * @brief `assert`-like macro for device code that persists in Release builds.
+ *
+ * This is effectively the same as the standard `assert` macro, except it will
+ * not be compiled out in Release builds, i.e., this macro will be present
+ * regardless of the state of `NDEBUG`.
+ *
+ * Relies on the `__PRETTY_FUNCTION__` macro which is specific to GCC and Clang.
+ *
+ *---------------------------------------------------------------------------**/
+#if defined(__CUDACC_ARCH__) && (defined(__clang__) || defined(__GNUC__))
+#define __ASSERT_STR_HELPER(x) #x
+#define device_assert(e)                                          \
```

Comment:
I would have thought it should be called "release_assert", since the only difference from "assert" is that it is not compiled out in release builds, not that it is device callable. CUDA already has device `assert()`.

Reply:
True, my reasoning is that it's only valid to be called from device code since `__assertfail` is a device intrinsic. 

Is `device_release_assert` too verbose?

Reply:
@jrhemstad : But it does the same thing a `release_assert()` would do on the host-side. We don't have `device_printf()`, for example. So, yes, too verbose.

I sometimes go with a different verb altogether, like "ensure" or "enforce" - so that people reading the code can't make any assumptions about release vs debug behavior. But `release_assert()` is ok too.